cpp/dynamic_community_detection_module/algorithm/dynamic_community_detection.cpp

```diff
@@ -0,0 +1,371 @@
+    if (P > max_P || (P == max_P && (std::int64_t)label < node_label)) {
+      max_P = P;
+      node_label = label;
+    }
+  }
+
+  return node_label;
+}
+
+std::unordered_map<std::uint64_t, std::int64_t> LabelRankT::AllLabels() {
+  std::unordered_map<std::uint64_t, std::int64_t> labels;
+
+  std::set<std::int64_t> labels_ordered;
+  std::map<std::int64_t, std::int64_t> lookup;
+
+  for (const auto [node, _] : label_Ps) {
+    labels_ordered.insert(NodeLabel(node));
+  }
+
+  int64_t label_i = 1;
+  for (const auto label : labels_ordered) {
+    if (label == -1) {  // no label found
+      lookup[label] = label;
+    } else {
+      lookup[label] = label_i;
+      label_i++;
+    }
+  }
+
+  for (const auto [node, _] : label_Ps) {
+    labels.insert({node, lookup[NodeLabel(node)]});
+  }
+
+  this->graph.reset();
+
+  return labels;
+}
+
+std::vector<std::uint64_t> LabelRankT::MostProbableLabels(
+    const std::uint64_t node_id) {
+  double max_P = 0;
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P > max_P) max_P = P;
+  }
+
+  std::vector<std::uint64_t> most_probable_labels = {};
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P == max_P) most_probable_labels.push_back(label);
+  }
+
+  return most_probable_labels;
+}
+/* #endregion */
+
+/* #region label_propagation_steps */
+bool LabelRankT::DistinctEnough(const std::uint64_t node_i_id,
+                                const double similarity_threshold) {
+  const std::vector<std::uint64_t> labels_i = MostProbableLabels(node_i_id);
+  std::uint64_t label_similarity = 0;
+
+  const auto in_neighbors_ids = InNeighborsMemgraphIDs(node_i_id);
+  for (const auto node_j_id : in_neighbors_ids) {
+    const std::vector<std::uint64_t> labels_j = MostProbableLabels(node_j_id);
+    if (IsSubset(labels_i, labels_j)) label_similarity++;
+  }
+
+  int node_i_in_degree = in_neighbors_ids.size();
+
+  return label_similarity <= node_i_in_degree * similarity_threshold;
+}
+
+std::unordered_map<std::uint64_t, double> LabelRankT::Propagate(
+    const std::uint64_t node_i_id) {
+  std::unordered_map<std::uint64_t, double> new_label_Ps_i;
+
+  // propagate own probabilities (handle self-loops)
+  for (const auto [label, P] : label_Ps[node_i_id]) {
+    new_label_Ps_i.insert({label, w_selfloop / sum_w[node_i_id] * P});
+  }
+
+  // propagate neighbors’ probabilities
+  for (const auto node_j_id : InNeighborsMemgraphIDs(node_i_id)) {
+    const double contribution =
+        GetTotalWeightBetween(node_j_id, node_i_id) / sum_w[node_i_id];
+    for (const auto [label, P] : label_Ps[node_j_id]) {
+      new_label_Ps_i[label] += contribution * P;
+    }
+  }
+
+  return new_label_Ps_i;
+}
+
+void LabelRankT::Inflate(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps,
+    const double exponent) {
+  double sum_Ps = 0;
+
+  for (const auto [label, _] : node_label_Ps) {
+    const double inflated_node_label_Ps = pow(node_label_Ps[label], exponent);
```

Comment:
As probabilities by definition are in the [0, 1] range, that is not the case. Anyhow, `std::pow` returns an appropriate value (+inf) in case an overflow occurs.